File: eclipse-plugin/com.atec.sourcetrace.eclipse/src/com/atec/sourcetrace/eclipse/core/TraceHttpClient.java
package com.atec.sourcetrace.eclipse.core;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

/**
 * HTTP client for Backend v2.6 using {@link HttpURLConnection}.
 * <p>
 * Uses fixed-length POST bodies and explicit {@code Content-Type: application/json}
 * so FastAPI receives a real JSON body (empty/missing body → HTTP 422
 * {@code type=missing, loc=["body"]}).
 */
public final class TraceHttpClient {

	public static final class HttpResult {
		public final int status;
		public final String body;
		public final String error;

		public HttpResult(int status, String body, String error) {
			this.status = status;
			this.body = body;
			this.error = error;
		}

		public boolean ok() {
			return error == null && status >= 200 && status < 300;
		}
	}

	public static final class EquipmentItem {
		public final int id;
		public final String name;

		public EquipmentItem(int id, String name) {
			this.id = id;
			this.name = name;
		}

		@Override
		public String toString() {
			return name + " (ID " + id + ")";
		}
	}

	public static final class RepoItem {
		public final int id;
		public final String name;
		public final String status;

		public RepoItem(int id, String name, String status) {
			this.id = id;
			this.name = name;
			this.status = status;
		}

		@Override
		public String toString() {
			return name + " (#" + id + ")" + (status != null ? " [" + status + "]" : "");
		}
	}

	private final int timeoutMs;

	public TraceHttpClient(Duration timeout) {
		Duration t = timeout == null ? Duration.ofSeconds(180) : timeout;
		this.timeoutMs = (int) Math.min(Integer.MAX_VALUE, Math.max(1000, t.toMillis()));
	}

	public TraceHttpClient() {
		this(Duration.ofSeconds(180));
	}

	public HttpResult get(String url) {
		HttpURLConnection conn = null;
		try {
			conn = open(url, "GET", false);
			conn.setRequestProperty("Accept", "application/json");
			int status = conn.getResponseCode();
			String body = readBody(conn, status);
			return new HttpResult(status, body, null);
		} catch (Exception e) {
			return new HttpResult(-1, null, e.getMessage() == null ? e.toString() : e.getMessage());
		} finally {
			if (conn != null) {
				conn.disconnect();
			}
		}
	}

	public HttpResult postJson(String url, String jsonBody) {
		HttpURLConnection conn = null;
		try {
			String payload = normalizeJsonBody(jsonBody);
			byte[] bytes = payload.getBytes(StandardCharsets.UTF_8);
			conn = open(url, "POST", true);
			conn.setRequestProperty("Accept", "application/json");
			conn.setRequestProperty("Content-Type", "application/json; charset=utf-8");
			conn.setFixedLengthStreamingMode(bytes.length);
			try (OutputStream os = conn.getOutputStream()) {
				os.write(bytes);
				os.flush();
			}
			int status = conn.getResponseCode();
			String body = readBody(conn, status);
			return new HttpResult(status, body, null);
		} catch (Exception e) {
			return new HttpResult(-1, null, e.getMessage() == null ? e.toString() : e.getMessage());
		} finally {
			if (conn != null) {
				conn.disconnect();
			}
		}
	}

	/** Never send blank — FastAPI treats missing/empty body as 422 body required. */
	public static String normalizeJsonBody(String jsonBody) {
		if (jsonBody == null || jsonBody.isBlank()) {
			return "{}";
		}
		return jsonBody;
	}

	private HttpURLConnection open(String url, String method, boolean doOutput) throws Exception {
		URL u = URI.create(url).toURL();
		HttpURLConnection conn = (HttpURLConnection) u.openConnection();
		conn.setConnectTimeout(Math.min(timeoutMs, 15_000));
		conn.setReadTimeout(timeoutMs);
		conn.setInstanceFollowRedirects(false);
		conn.setRequestMethod(method);
		conn.setDoInput(true);
		conn.setDoOutput(doOutput);
		conn.setUseCaches(false);
		return conn;
	}

	private static String readBody(HttpURLConnection conn, int status) throws Exception {
		InputStream in = status >= 400 ? conn.getErrorStream() : conn.getInputStream();
		if (in == null) {
			in = conn.getInputStream();
		}
		if (in == null) {
			return "";
		}
		try (InputStream stream = in; ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
			byte[] buf = new byte[4096];
			int n;
			while ((n = stream.read(buf)) >= 0) {
				bos.write(buf, 0, n);
			}
			return bos.toString(StandardCharsets.UTF_8);
		}
	}

	public HttpResult health(String serverUrl) {
		return get(ServerUrlUtil.buildApiUrl(serverUrl, ApiPaths.HEALTH));
	}

	public List<EquipmentItem> listEquipment(String serverUrl) throws Exception {
		HttpResult r = get(ServerUrlUtil.buildApiUrl(serverUrl, ApiPaths.EQUIPMENT_LIST));
		if (!r.ok()) {
			throw new Exception(formatHttpError("장비 목록 조회", r));
		}
		List<EquipmentItem> out = new ArrayList<>();
		for (String obj : JsonUtil.getObjectArrayItems(r.body, null)) {
			Integer id = JsonUtil.getIntField(obj, "id");
			String name = JsonUtil.getStringField(obj, "name");
			if (id != null) {
				out.add(new EquipmentItem(id.intValue(), name == null ? "" : name));
			}
		}
		return out;
	}

	public List<RepoItem> listRepositories(String serverUrl, int equipmentId) throws Exception {
		HttpResult r = get(ServerUrlUtil.buildApiUrl(serverUrl, ApiPaths.equipmentRepositories(equipmentId)));
		if (!r.ok()) {
			throw new Exception(formatHttpError("Repository 목록 조회", r));
		}
		List<RepoItem> out = new ArrayList<>();
		for (String obj : JsonUtil.getObjectArrayItems(r.body, null)) {
			Integer id = JsonUtil.getIntField(obj, "id");
			String name = JsonUtil.getStringField(obj, "name");
			String status = JsonUtil.getStringField(obj, "status");
			if (id != null) {
				out.add(new RepoItem(id.intValue(), name == null ? "" : name, status));
			}
		}
		return out;
	}

	public HttpResult postReport(String serverUrl, String jsonBody) {
		return postJson(ServerUrlUtil.buildApiUrl(serverUrl, ApiPaths.TRACE_REPORT), jsonBody);
	}

	public HttpResult postSelection(String serverUrl, String jsonBody) {
		return postJson(ServerUrlUtil.buildApiUrl(serverUrl, ApiPaths.TRACE_SELECTION), jsonBody);
	}

	public static String formatHttpError(String action, HttpResult r) {
		if (r.error != null) {
			return action + " 실패: 서버에 연결할 수 없습니다. (" + r.error + ")";
		}
		if (r.status == 422) {
			return FastApiErrorParser.formatUserMessage(action, r.status, r.body);
		}
		String detail = r.body == null ? "" : r.body;
		String msg = JsonUtil.getStringField(detail, "detail");
		if (msg == null || msg.isEmpty()) {
			msg = JsonUtil.getStringField(detail, "content");
		}
		if (msg == null || msg.isEmpty()) {
			msg = detail.length() > 300 ? detail.substring(0, 300) : detail;
		}
		return action + "에 실패했습니다. (HTTP " + r.status + ")\n" + msg;
	}

	public static String userFacingConnectionError(String detail) {
		return "Backend 서버에 연결할 수 없습니다. Source Trace 서버 URL 설정을 확인하세요."
				+ (detail != null && !detail.isEmpty() ? "\n원인: " + detail : "");
	}
}

File: visualstudio-extension/src/Atec.SourceTrace.Core/TraceHttpClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;

namespace Atec.SourceTrace.Core;

/// <summary>
/// HTTP client using HttpWebRequest with fixed-length UTF-8 POST bodies
/// (prevents FastAPI 422 missing body).
/// </summary>
public sealed class TraceHttpClient
{
    public sealed class HttpResult
    {
        public int Status { get; }
        public string? Body { get; }
        public string? Error { get; }

        public HttpResult(int status, string? body, string? error)
        {
            Status = status;
            Body = body;
            Error = error;
        }

        public bool Ok => Error == null && Status >= 200 && Status < 300;
    }

    public sealed class EquipmentItem
    {
        public int Id { get; }
        public string Name { get; }

        public EquipmentItem(int id, string name)
        {
            Id = id;
            Name = name;
        }

        public override string ToString() => $"{Name} (ID {Id})";
    }

    public sealed class RepoItem
    {
        public int Id { get; }
        public string Name { get; }
        public string? Status { get; }

        public RepoItem(int id, string name, string? status)
        {
            Id = id;
            Name = name;
            Status = status;
        }

        public override string ToString() =>
            $"{Name} (#{Id})" + (Status != null ? $" [{Status}]" : string.Empty);
    }

    private readonly int _timeoutMs;

    public TraceHttpClient(TimeSpan? timeout = null)
    {
        var t = timeout ?? TimeSpan.FromSeconds(180);
        _timeoutMs = (int)Math.Min(int.MaxValue, Math.Max(1000, t.TotalMilliseconds));
    }

    public HttpResult Get(string url) => Send(url, "GET", null);

    public HttpResult PostJson(string url, string? jsonBody)
    {
        var payload = NormalizeJsonBody(jsonBody);
        return Send(url, "POST", payload);
    }

    public static string NormalizeJsonBody(string? jsonBody)
    {
        return string.IsNullOrWhiteSpace(jsonBody) ? "{}" : jsonBody;
    }

    private HttpResult Send(string url, string method, string? jsonPayload)
    {
        try
        {
            var request = (HttpWebRequest)WebRequest.Create(url);
            request.Method = method;
            request.Accept = "application/json";
            request.Timeout = _timeoutMs;
            request.ReadWriteTimeout = _timeoutMs;
            request.AllowAutoRedirect = false;

            if (method == "POST")
            {
                var bytes = Encoding.UTF8.GetBytes(jsonPayload ?? "{}");
                request.ContentType = "application/json; charset=utf-8";
                request.ContentLength = bytes.Length;
                using (var stream = request.GetRequestStream())
                {
                    stream.Write(bytes, 0, bytes.Length);
                    stream.Flush();
                }
            }

            using var response = (HttpWebResponse)request.GetResponse();
            var body = ReadResponseBody(response);
            return new HttpResult((int)response.StatusCode, body, null);
        }
        catch (WebException ex) when (ex.Response is HttpWebResponse err)
        {
            var body = ReadResponseBody(err);
            return new HttpResult((int)err.StatusCode, body, null);
        }
        catch (Exception ex)
        {
            return new HttpResult(-1, null, ex.Message);
        }
    }

    private static string ReadResponseBody(HttpWebResponse response)
    {
        var stream = response.GetResponseStream();
        if (stream == null)
        {
            return string.Empty;
        }

        using var reader = new StreamReader(stream, Encoding.UTF8);
        return reader.ReadToEnd();
    }

    public HttpResult Health(string serverUrl) =>
        Get(ServerUrlUtil.BuildApiUrl(serverUrl, ApiPaths.Health));

    public List<EquipmentItem> ListEquipment(string serverUrl)
    {
        var r = Get(ServerUrlUtil.BuildApiUrl(serverUrl, ApiPaths.EquipmentList));
        if (!r.Ok)
        {
            throw new Exception(FormatHttpError("장비 목록 조회", r));
        }

        var outList = new List<EquipmentItem>();
        foreach (var obj in JsonUtil.GetObjectArrayItems(r.Body, null))
        {
            var id = JsonUtil.GetIntField(obj, "id");
            var name = JsonUtil.GetStringField(obj, "name") ?? string.Empty;
            if (id.HasValue)
            {
                outList.Add(new EquipmentItem(id.Value, name));
            }
        }

        return outList;
    }

    public List<RepoItem> ListRepositories(string serverUrl, int equipmentId)
    {
        var r = Get(ServerUrlUtil.BuildApiUrl(serverUrl, ApiPaths.EquipmentRepositories(equipmentId)));
        if (!r.Ok)
        {
            throw new Exception(FormatHttpError("Repository 목록 조회", r));
        }

        var outList = new List<RepoItem>();
        foreach (var obj in JsonUtil.GetObjectArrayItems(r.Body, null))
        {
            var id = JsonUtil.GetIntField(obj, "id");
            var name = JsonUtil.GetStringField(obj, "name") ?? string.Empty;
            var status = JsonUtil.GetStringField(obj, "status");
            if (id.HasValue)
            {
                outList.Add(new RepoItem(id.Value, name, status));
            }
        }

        return outList;
    }

    public HttpResult PostReport(string serverUrl, string jsonBody) =>
        PostJson(ServerUrlUtil.BuildApiUrl(serverUrl, ApiPaths.TraceReport), jsonBody);

    public HttpResult PostSelection(string serverUrl, string jsonBody) =>
        PostJson(ServerUrlUtil.BuildApiUrl(serverUrl, ApiPaths.TraceSelection), jsonBody);

    public static string FormatHttpError(string action, HttpResult r)
    {
        if (r.Error != null)
        {
            return action + " 실패: 서버에 연결할 수 없습니다. (" + r.Error + ")";
        }

        if (r.Status == 422)
        {
            return FastApiErrorParser.FormatUserMessage(action, r.Status, r.Body);
        }

        var detail = r.Body ?? string.Empty;
        var msg = JsonUtil.GetStringField(detail, "detail")
                  ?? JsonUtil.GetStringField(detail, "content")
                  ?? (detail.Length > 300 ? detail.Substring(0, 300) : detail);
        return action + "에 실패했습니다. (HTTP " + r.Status + ")\n" + msg;
    }

    public static string UserFacingConnectionError(string? detail) =>
        "Backend 서버에 연결할 수 없습니다. Source Trace 서버 URL 설정을 확인하세요."
        + (string.IsNullOrEmpty(detail) ? string.Empty : "\n원인: " + detail);
}
